src/Mapbender/CoreBundle/Resources/public/mapbender.element.legend.js

```diff
@@ -1,515 +1,478 @@
-            if(this.options.checkGraphic) {
-                this._createCheckedLegendHtml(sources, 0, 0, "", "", $.proxy(self._createLegend, self));
-            } else {
-                this._createLegend(this._createLegendHtml(sources));
-            }
-            sources = this.model.getSources();
-            for(var i = 0; i < sources.length; i++) {
-                this._checkLayers(sources[i]);
-            }
+        render: function() {
+            var widget = this;
+            var sources = widget._getSources();
+            var html = $("<ul/>");
+            _.each(sources, function(source) {
+                html.append(widget._createLayerHtml(source));
+            });
+            return html;
+        },
+
+        /**
+         * On open handler
+         */
+        open: function(callback) {
+            var widget = this;
 
+            widget.callback = callback;
+
+            if(widget.isPopUpDialog) {
+                widget.popup(widget.htmlContainer.html());
+            }
         },
-        close: function(){
-            if(this.options.elementType === "dialog") {
-                if(this.popup) {
-                    if(this.popup.$element)
-                        this.popup.destroy();
-                    this.popup = null;
+
+        /**
+         * On close
+         */
+        close: function() {
+            var widget = this;
+            if(widget.isPopUpDialog) {
+                if(widget.popup) {
+                    if(widget.popupWindow.$element) widget.popupWindow.destroy();
+                    widget.popupWindow = null;
                 }
 
             }
-            if(this.callback) {
-                this.callback.call();
-                this.callback = null;
+            if(widget.callback) {
+                widget.callback.call();
+                widget.callback = null;
             }
         },
+
         /**
-         *
+         * On ready handler
          */
-        ready: function(callback){
-            if(this.readyState === true) {
-                callback();
+        ready: function(callback) {
+            var widget = this;
+            if(widget.readyState) {
+                if(typeof(callback ) === 'function') {
+                    callback();
+                }
             } else {
-                this.readyCallbacks.push(callback);
+                widget.readyCallbacks.push(callback);
             }
         },
+
         /**
-         *
+         * On ready handler
          */
-        _ready: function(){
-            for(callback in this.readyCallbacks) {
-                callback();
-                delete(this.readyCallbacks[callback]);
+        _ready: function() {
+            var widget = this;
+            var readyCallbacks = widget.readyCallbacks;
+            for (var k in readyCallbacks) {
+                var readyCallback = readyCallbacks[k];
+                if(typeof(readyCallback ) === 'function') {
```

Comment:
Instead of proofing that it is a function we can add a `try catch` before we run the loop, or the forEach function. With this way, we can handle errors from the callbacks also...